### PR TITLE
fix(server): return 404 if get backing image and backup not found

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: [longhorn/longhorn#6266](https://github.com/longhorn/longhorn/issues/6266)

### CODE REVIEW

File: api/backingimage.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import (
 	"net/http"
 
 	"github.com/gorilla/mux"
+	"github.com/longhorn/longhorn-manager/datastore"
```
Comment (shuo-wu) [inline api/backingimage.go:8]: NIT: re-arrange this.

File: api/backup.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"net/http"
 
 	"github.com/gorilla/mux"
+	"github.com/longhorn/longhorn-manager/datastore"
```
Comment (shuo-wu) [inline api/backup.go:7]: NIT: re-arrange this.

**Review by shuo-wu — APPROVED** [review]
Do we need to handle volume/backupvolume for CSI API `DeleteSnapshot`?
The idea LGTM.

File: api/backup.go (hunk 2 above)
Comment (ChanYiLin) [reply api/backup.go]: Fixed, thanks!

File: api/backingimage.go (hunk 1 above)
Comment (ChanYiLin) [reply api/backingimage.go]: Fixed.

**Comment by ChanYiLin** [comment]
VolumeSnapshot has 3 types
- backing image
- snapshot
- backup

When deleting
- csi: try to get the CR first, it it doesn't exist, then pass
- client: check if 404, return nil
- server: return 404, if it cannot get the CR from datastore

**Review by innobead — DISMISSED** [review]
LGTM.

**Comment by innobead** [comment]
@shuo-wu Please help review this.

**Comment by ChanYiLin** [comment]
@innobead 
I am running the e2e test, and it failed at backup snapshot
Let me check for it

**Comment by innobead** [comment]
NP. Passing the test is the criteria before merging.

**Comment by ChanYiLin** [comment]
weird, it passed the test in my environment
need some time to investigate

**Comment by ChanYiLin** [comment]
I know the reason why
Because I use `/v1/backups/XXX` in our generated client to try to get the backup before deleting it 
Turns out, we don't provide such API in our server
So csi thought there was no such backup and pass the deletion, and failed when checking if the backup still existed in the python test.

**Comment by ChanYiLin** [comment]
We don't have to check by get the backup or snapshot before deletion
because in the delete api we already pass the deletion if it is not found

https://github.com/longhorn/longhorn-manager/blob/75d2e77c539bd8a635140b87e81779b7d049787b/api/backup.go#L125-L134

so removed the check

**Comment by ChanYiLin** [comment]
e2e running
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4550/console

**Comment by shuo-wu** [comment]
> We don't have to check by getting the backup or snapshot before deletion
because in the delete api we already pass the deletion if it is not found
[longhorn-manager/api/backup.go](https://github.com/longhorn/longhorn-manager/blob/75d2e77c539bd8a635140b87e81779b7d049787b/api/backup.go#L125-L134)
so removed the check

Sorry... Does the PR contain the removal (of getting backing image/backup volume before requesting deletion)...? BTW, for the snapshot deletion, getting volume is unavoidable. We still need to make sure getting a nil volume should be handled as expected.

**Comment by ChanYiLin** [comment]
Hi @shuo-wu 
No, this PR maintain the same as before
And only add 404 not found header in server side
For csi part, I remove all the change I made since the deletion api for `backup` and `snapshot` already check if it exists

**Comment by ChanYiLin** [comment]
e2e tests passed
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/4550/

**Comment by shuo-wu** [comment]
OK. I was thinking you directly remove the backup volume/backing image get calls from the CSI `SnapshotDelete` implementation. But this is an alternative solution for this issue while making the implementation simpler. WDTY?

**Comment by ChanYiLin** [comment]
I think if from server side it already check if it exists before delete it
We don't need to do it in client side again 
and if we do it in client side, we have to make sure every client call has checked if it exists? 
I would say it adds the burden to the client side.

**Comment by innobead** [comment]
So do we need to do any other code change here?

**Comment by ChanYiLin** [comment]
> So do we need to do any other code change here?

No, I have tested it manually and through e2e
Three types of volumesnapshot can be successfully deleted in any case

**Comment by shuo-wu** [comment]
I mean, for backing image cleanup, since there is no parent resource, there is no need to invoke `cs.apiClient.BackingImage.ById(backingImageParameters[longhorn.BackingImageParameterName])` before deletion. We can directly call `cs.apiClient.BackingImage.Delete(backingImage)`

**Comment by ChanYiLin** [comment]
@shuo-wu
But in csi we don't have longhorn CR clientset only have longhorn api client
so we manipulate the resources through api client
and for BackingImage delete api, it needs BackingImage object

Even we delete the BackingImage directly, if it doesn't exist, we still need to handle not found error
I don't think the process will be different
- get -> if found -> delete
- delete -> if not found error then pass

And besides this issue, we still need to add 404 not found in the server api for other cases, maybe there are some places in our code also use 404 to determine if api has not found error.

**Review by shuo-wu — APPROVED** [review]

**Comment by innobead** [comment]
@mergify backport v1.5.x

**Pull request merged**